sandbox/cgroup/tracking.go

```diff
@@ -127,11 +128,22 @@ tryAgain:
 	//
 	// Verify the effective tracking cgroup and check that our scope name is
 	// contained therein.
-	path, err := cgroupProcessPathInTrackingCgroup(pid)
-	if err != nil {
-		return err
+	hasTracking := false
+	start := time.Now()
+	for tries := 0; tries < 100; tries++ {
+		path, err := cgroupProcessPathInTrackingCgroup(pid)
+		if err != nil {
+			return err
+		}
+		if strings.HasSuffix(path, unitName) {
+			hasTracking = true
+			break
+		}
+		time.Sleep(1 * time.Millisecond)
```

Comment:
(niptick) 1 Millisecond seems rather aggressive here, maybe 5 or 10 ?

Reply:
This is in the execution path and introduces additional latency. I think we should keep the delay reasonably low. On my system, I have seen the first check to be unsuccessful almost always, so at least one time.Sleep() was executed.

Reply:
I think 1 millisecond is fine here, I think the latency is more important than wasted CPU cycles since this is the hot path for every snap app we execute

Reply:
we had problems in the past though with too aggressive loops on low powered cpus,  how long does cgroupProcessPathInTrackingCgroup itself here take on something like a RPi or similar?

Reply:
with this benchmark test:
```go
var result string
var err error

func BenchmarkProcessPathInTrackingCgroup(b *testing.B) {

	d, _ := ioutil.TempDir("", "")
	restore := cgroup.MockFsRootPath(d)
	defer restore()

	restore = cgroup.MockVersion(cgroup.V2, nil)
	defer restore()

	f := filepath.Join(d, "proc", "1234", "cgroup")
	os.MkdirAll(filepath.Dir(f), 0755)

	const noise = `12:cpuset:/
11:rdma:/
10:blkio:/
9:freezer:/
8:cpu,cpuacct:/
7:perf_event:/
6:net_cls,net_prio:/
5:devices:/user.slice
4:hugetlb:/
3:memory:/user.slice/user-1000.slice/user@1000.service
2:pids:/user.slice/user-1000.slice/user@1000.service
`

	ioutil.WriteFile(f, []byte(noise+"1:name=systemd:/bar"), 0644)

	// vars to prevent compiler optimization
	var r string
	var e error

	// run the func b.N times
	for n := 0; n < b.N; n++ {
		r, e = cgroup.ProcessPathInTrackingCgroup(1234)
	}

	result = r
	err = e
}
```

on amd64 I see:
```
goos: linux
goarch: amd64
pkg: github.com/snapcore/snapd/sandbox/cgroup
BenchmarkProcessPathInTrackingCgroup-32            75156             15310 ns/op
PASS
ok      github.com/snapcore/snapd/sandbox/cgroup        1.320s
```

and on my raspi 3, which is single core I see:
```
goos: linux
goarch: arm64
pkg: github.com/snapcore/snapd/sandbox/cgroup
BenchmarkProcessPathInTrackingCgroup-4   	   11442	    102602 ns/op
PASS
ok  	github.com/snapcore/snapd/sandbox/cgroup	2.060s
```

so it's taking about 100 ms to execute on arm, but 15 ms to execute on amd64.

Reply:
so the next question is whether those 100ms/15ms are mostly cpu or i/o wait, if the former we need to be careful, if the latter it doesn't matter much

Reply:
I can look into implementing the actual dbus iteraction, in which case we won't have busy waiting anymore.

Reply:
Note that 102602 ns/op means about 100us, not 100ms.

Reply:
heh, indeed

Reply:
I've made some measurements on Pi4. When the system is idle it takes ~4 tries (~4ms) to observe the process in the right cgroup if the app is run by root (i.e. communicating with system instance of systemd), or ~5 (~5.2ms) when the app is run by a regular user. I understand that the difference comes from the fact that the process is moved to the right cgroup by the system instance (systemd --user talks to systemd to set it up).
Under load (`stress --cpu 4`), the wait is shorter, as the `snap` process is competing for the CPU and takes longer to reach the point where we do the check.